pkg/ingester/ingester_v2.go

```diff
@@ -78,20 +86,30 @@ func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*clien
 		return nil, err
 	}
 
+	// Ensure the ingester shutdown procedure hasn't started
+	i.userStatesMtx.RLock()
+	stopped := i.stopped
+	i.userStatesMtx.RUnlock()
+
+	if stopped {
+		return nil, fmt.Errorf("ingester stopping")
+	}
+
+	// Keep track of in-flight requests, in order to safely start blocks transfer
```

Comment:
I believe there's a race condition here. If the ingester isn't stopped, gets passed the check, and gets stopped before incrementing the `inflightWriteReqs` we can start a transfer and a write can that continue on.

Reply:
Right! I think the safest way would be incrementing the `sync.WaitGroup` within the lock. What's about now?

Reply:
hmm I think I was wrong and that doesn't fix it. Since inflightwritereqs is being waited on outside the lock I believe the race condition still exists.

Reply:
Here we check if the flag `stopped` is set. If not, we do proceed increasing the `inflightWriteReqs` within the lock wrapping the `stopped`.

At ingester shutdown, when the `v2TransferOut()` will be called, the `stopped` flag is already set to `true` (it uses a write lock), which means that when the `stopped` flag is set to `true` there's the guarantee no other routine will be within the read lock used in `v2Push()`. Then, outside the lock, down in the execution path, we do wait on `inflightWriteReqs` but there's already the guarantee it won't be incremented because the `stopped` flag is already set to `true`.

Am I missing anything?

Reply:
okay you're right, I think I just got lost in my head. Thanks for fixing it!